Simplify `computeContainerBoundingBoxes` by introducing helper function

Ahmad-S792 opened the pull request:
#### f7086dbdee393ec8a768399d296b1c8bf7fddc98
<pre>
Simplify `computeContainerBoundingBoxes` by introducing helper function

<a href="https://bugs.webkit.org/show_bug.cgi?id=290124">https://bugs.webkit.org/show_bug.cgi?id=290124</a>
<a href="https://rdar.apple.com/147524534">rdar://147524534</a>

Reviewed by Darin Adler.

This patch introduces new helper function `hasValidBoundingBoxForContainer`,
which is to manage bounding box for `bbox` as single source and this is
follow-up to 292411@main, where we merged below Blink commit [1] but not fully.

[1] <a href="https://chromium.googlesource.com/chromium/src/+/e47e6e98afe2de2d737ac7827e9309430e68b60b">https://chromium.googlesource.com/chromium/src/+/e47e6e98afe2de2d737ac7827e9309430e68b60b</a>

* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::hasValidBoundingBoxForContainer):
(WebCore::SVGRenderSupport::computeContainerBoundingBoxes):

Canonical link: <a href="https://commits.webkit.org/295439@main">https://commits.webkit.org/295439@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/69861b82ec307c52f0a960b7416a03bf28304b97

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/105095 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24809 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/15232 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/110312 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/55775 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/107136 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/25230 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/33353 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79815 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/108101 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/19654 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94862 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60122 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/19410 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12937 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/55152 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/89121 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12982 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/112839 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/32260 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23756 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88895 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/32626 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/91081 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88525 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22568 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33419 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/11210 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/27660 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/32183 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/37559 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31976 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/35317 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/33535 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler reviewed (changes requested):
I don’t think it’s appropriate to title a pull request that changes SVG behavior as "simplify with helper function". We need to title with the behavior change, and clarify why this is a good progression and desirable change. Alternatively, could separate the behavior change from the refactoring.

darinadler commented on Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
```diff
@@ -141,25 +141,33 @@ LayoutRepainter::CheckForRepaint SVGRenderSupport::checkForSVGRepaintDuringLayou
 {
-    ContainerBoundingBoxes result;
+    if (auto* svgShape = dynamicDowncast<LegacyRenderSVGShape>(object))
+        return svgShape && !svgShape->isRenderingDisabled();
```
I don’t think our local variable name has to be this specific. Can just be a sensible word:
```suggestion
    if (auto* shape = dynamicDowncast<LegacyRenderSVGShape>(object))
        return shape && !shape->isRenderingDisabled();
```

Ahmad-S792 replied:
Done.

darinadler commented on Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
```diff
@@ -141,25 +141,33 @@ LayoutRepainter::CheckForRepaint SVGRenderSupport::checkForSVGRepaintDuringLayou
-        if (auto* shape = dynamicDowncast<LegacyRenderSVGShape>(current.ptr()); shape && shape->isRenderingDisabled())
-            continue;
+    if (auto* hiddenContainer = dynamicDowncast<LegacyRenderSVGContainer>(object))
+        return hiddenContainer && !hiddenContainer->isLegacyRenderSVGHiddenContainer();
```
I don’t think our local variable name has to be this specific. Can just be a sensible word. Also, I think this is making an object that is a non-hidden container be in a local variable named hidden container.
```suggestion
    if (auto* container = dynamicDowncast<LegacyRenderSVGContainer>(object))
        return container && !container->isLegacyRenderSVGHiddenContainer();
```

Ahmad-S792 replied:
Done.

darinadler commented on Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
```diff
@@ -141,25 +141,33 @@ LayoutRepainter::CheckForRepaint SVGRenderSupport::checkForSVGRepaintDuringLayou
-        if (auto* image = dynamicDowncast<LegacyRenderSVGImage>(current.ptr()); (image && !image->isObjectBoundingBoxValid()))
-            continue;
+    if (auto* svgImage = dynamicDowncast<LegacyRenderSVGImage>(object))
+        return svgImage && svgImage->isObjectBoundingBoxValid();
```
I don’t think our local variable name has to be this specific. Can just be a sensible word:
```suggestion
    if (auto* image = dynamicDowncast<LegacyRenderSVGImage>(object))
        return image && !image->isObjectBoundingBoxValid();
```

Ahmad-S792 replied:
Done.

darinadler commented on Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
```diff
@@ -141,25 +141,33 @@ LayoutRepainter::CheckForRepaint SVGRenderSupport::checkForSVGRepaintDuringLayou
+
+    for (CheckedRef current : childrenOfType<RenderObject>(container)) {
+        // Don't include elements that are not rendered in the union.
+        if (!hasValidBoundingBoxForContainer(current))
```
It’s a bad sign that the comment says “elements that are not rendered” but the function name says “has valid bounding box”. There’s no reason to not name the code the same way we’d describe it to a person. Then we wouldn’t need a comment because the code would say the same thing as the comment! So either the comment should change to match the function name and then perhaps be removed, or the function name should change to match the comment and then perhaps the comment would be removed.

If we keep the comment, it should explain _why_ the code does what it does, not just repeat what the code does with different words.

Ahmad-S792 replied:
I removed this comment but added link to specification which describes what we are doing in `hasValidBoundingBoxForContainer` function to have valid bounding box for container. It can make it easier to understand this code.

shallawa commented on Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
```diff
@@ -141,25 +141,33 @@ LayoutRepainter::CheckForRepaint SVGRenderSupport::checkForSVGRepaintDuringLayou
-        if (current->isLegacyRenderSVGHiddenContainer())
-            continue;
+    if (auto* text = dynamicDowncast<RenderSVGText>(object))
+        return text && text->isObjectBoundingBoxValid();
```
Why do we have to check `text` again in the return statement? Is not it enough to say:
```
   if (auto* text = dynamicDowncast<RenderSVGText>(object))
        return text->isObjectBoundingBoxValid();
```

Ahmad-S792 replied:
@shallawa - done. Sorry for being late to rebase and update this.

darinadler commented on Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
```diff
@@ -142,25 +142,33 @@ LayoutRepainter::CheckForRepaint SVGRenderSupport::checkForSVGRepaintDuringLayou
-        if (auto* shape = dynamicDowncast<LegacyRenderSVGShape>(current.ptr()); shape && shape->isRenderingDisabled())
-            continue;
+    if (auto* container = dynamicDowncast<LegacyRenderSVGContainer>(object))
+        return container && !container->isLegacyRenderSVGHiddenContainer();
```
```suggestion
        return !container->isLegacyRenderSVGHiddenContainer();
```

Ahmad-S792 replied:
Done for `shape` as well.

Ahmad-S792 committed:
Simplify `computeContainerBoundingBoxes` by introducing helper function

https://bugs.webkit.org/show_bug.cgi?id=290124
rdar://147524534

Reviewed by Darin Adler.

This patch introduces new helper function `hasValidBoundingBoxForContainer`,
which is to manage bounding box for `bbox` as single source and this is
follow-up to 292411@main, where we merged below Blink commit [1] but not fully.

[1] https://chromium.googlesource.com/chromium/src/+/e47e6e98afe2de2d737ac7827e9309430e68b60b

* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::hasValidBoundingBoxForContainer):
(WebCore::SVGRenderSupport::computeContainerBoundingBoxes):

Canonical link: https://commits.webkit.org/295439@main
